cmd/krew/cmd/update.go

```diff
@@ -41,12 +57,147 @@ Remarks:
+	}
+
+	return plugins, nil
+}
+
+func retrievePluginNameVersionMap(names []string) map[string]string {
+	m := make(map[string]string, len(names))
+	for _, n := range names {
+		plugin, err := indexscanner.LoadPluginByName(paths.IndexPluginsPath(), n)
+		if err != nil {
+			continue
+		}
+
+		m[n] = plugin.Spec.Version
+	}
+
+	return m
+}
+
+func retrieveInstalledPluginMap() (map[string]bool, error) {
+	plugins, err := installation.ListInstalledPlugins(paths.InstallReceiptsPath())
+	if err != nil {
+		return map[string]bool{}, err
+	}
+	m := make(map[string]bool, len(plugins))
+	for name := range plugins {
+		m[name] = true
+	}
+
+	return m, nil
+}
+
+func filterAndSortUpdatedPlugins(old, updated map[string]string, installed map[string]bool) ([]newPlugin, []newVersionPlugin) {
+	var newPluginList []newPlugin
+	var newVersionList []newVersionPlugin
+
+	for name, version := range updated {
+		oldVersion, ok := old[name]
+		if !ok {
+			newPluginList = append(newPluginList, newPlugin{
+				name:    name,
+				version: version,
+			})
+			continue
+		}
+
+		if version != oldVersion {
+			_, installed := installed[name]
+			newVersionList = append(newVersionList, newVersionPlugin{
+				name:       name,
+				installed:  installed,
+				oldVersion: oldVersion,
+				newVersion: version,
+			})
+			continue
+		}
+	}
+
+	sort.Slice(newPluginList, func(i, j int) bool {
+		return newPluginList[i].name < newPluginList[j].name
+	})
+
+	sort.Slice(newVersionList, func(i, j int) bool {
+		if newVersionList[i].installed && !newVersionList[j].installed {
+			return true
+		}
+
+		if !newVersionList[i].installed && newVersionList[j].installed {
+			return false
+		}
+
+		return newVersionList[i].name < newVersionList[j].name
+	})
+
+	return newPluginList, newVersionList
+}
+
+func showUpdatedPlugins(newPluginList []newPlugin, newVersionList []newVersionPlugin) {
+	if len(newPluginList) > 0 {
+		fmt.Fprintln(os.Stderr, "  New plugins available:")
+		for _, np := range newPluginList {
+			fmt.Fprintf(os.Stderr, "    * %s %s\n", np.name, np.version)
+		}
+	}
+
+	if len(newVersionList) > 0 {
+		fmt.Fprintln(os.Stderr, "  The following plugins have new version:")
+		for _, np := range newVersionList {
+			if np.installed {
+				fmt.Fprintf(os.Stderr, "    * %s %s -> %s (!)\n", np.name, np.oldVersion, np.newVersion)
+				continue
+			}
+			fmt.Fprintf(os.Stderr, "    * %s %s -> %s\n", np.name, np.oldVersion, np.newVersion)
+		}
+	}
+}
+
 func ensureIndexUpdated(_ *cobra.Command, _ []string) error {
+	updatedPlugins, err := retrieveUpdatedPluginList()
+	if err != nil {
+		return errors.Wrap(err, "failed to load the list of updated plugins from the index")
+	}
+
+	var oldMap map[string]string
+	if len(updatedPlugins) > 0 {
+		oldMap = retrievePluginNameVersionMap(updatedPlugins)
+	}
+
 	klog.V(1).Infof("Updating the local copy of plugin index (%s)", paths.IndexPath())
 	if err := gitutil.EnsureUpdated(constants.IndexURI, paths.IndexPath()); err != nil {
```

Comment:
`git fetch` will be executed twice. Here and in line 170. As this involves talking to a remote this is quite slow, and I would like to avoid the second one if possible.